ДЗ 13 спринт

HUEY-news opened the pull request:
**Критерии успешного выполнения задания**:

**1. Переход на экран «Аудиоплеер»**
- Если пользователь находится на экране «Поиск» и видит список треков, 
то при нажатии на элемент списка пользователь перенаправляется на экран «Аудиоплеер».
- Когда пользователь находится на экране «Поиск» и видит историю поиска,
то при нажатии на элемент списка пользователь перенаправляется на экран «Аудиоплеер».
- Если приложение было переведено в фоновый режим на экране «Аудиоплеер», 
то после восстановления приложения из фонового режима пользователь видит экран «Аудиоплеер».

**2. Переход на предыдущий экран**:
- Нажатие на системную кнопку «Назад» или применение аналогичного жеста 
возвращает пользователя на предыдущий экран.

**3. UI экрана аудиоплеера**:

Когда пользователь находится на экране аудиоплеера, то он видит следующие UI-элементы:
- Кнопку «Назад» (нажатие на неё эквивалентно нажатию системной кнопки «Назад»),
- Область для отображения обложки плеера,
- Название трека (trackName),
- Имя исполнителя (artistName),
- Название альбома (collectionName) (если есть),
- Год релиза трека (releaseDate),
- Жанр трека (primaryGenreName),
- Страна исполнителя (country),
- Продолжительность трека в формате mm:ss (trackTimeMills),
- Кнопка добавления в плейлист,
- Кнопка добавления в избранное,
- Кнопка контроля воспроизведения,
- Прогресс воспроизведения (время под кнопкой контроля воспроизведения в формате mm:ss).

Кроме этих полей, на экране будет отображаться обложка альбома.

**4. Плейсхолдер обложки в плеере**:

Если у трека нет изображения обложки, то на экране пользователь должен видеть изображение-заглушку.
Если у трека есть изображение обложки, но оно ещё не загружено, 
то пользователь должен видеть изображение-заглушку.

**5. Изображение обложки в плеере**:

- Если у трека есть изображение обложки и оно загружено приложением, 
то пользователь должен видеть это изображение в области для отображения обложки на экране аудиоплеера.
- В качестве изображения обложки следует использовать изображение, 
- находящееся по ссылке параметра artworkUrl100 с заменой последней части ссылки на 512x512bb.jpg.
- Если у трека есть изображение обложки, но оно ещё не загружено, 
то пользователь также должен видеть изображение-заглушку в области отображения обложки.
- У нас есть ссылка на обложку, но по этой ссылке изображение приходит в маленьком разрешении.
Это нам подходит для списка, но для аудиоплеера нам нужно качество лучше.

Вот так выглядит ссылка на обложку: 
https://is5-ssl.mzstatic.com/image/thumb/Music125/v4/a0/4d/c4/a04dc484-03cc-02aa-fa82-5334fcb4bc16/18UMGIM24878.rgb.jpg/100x100bb.jpg.
Она заканчивается на 100x100bb.jpg, и если мы поменяем конец ссылки на 512x512bb.jpg,
 то получим более качественное изображение.

HUEY-news committed:
1. Начат набросок экрана плеера.
2. Кнопки переделаны через атрибуты темы.

HUEY-news committed:
1. Добавлены строковые ресурсы информации о треке.
2. Заголовки track_info_title добавлены в макет.
3. Добавлен барьер в макет к заголовкам.
4. Завершён макет плеера.

HUEY-news committed:
1. Поле года сдвигается, если поля альбома нет.
2. Добавлены недостающие поля в модель трека.

HUEY-news committed:
1. Плеер теперь активити.
2. Избавился от пары ненужных туду.
3. В манифесте к плееру привязана тема.

HUEY-news committed:
1. Реализован переход по нажатию в поиске.
2. Реализована функция кнопки "назад".
3. Экран плеера заполняется с файла трека.

HUEY-news committed:
1. При нажатии на элемент списка пользователь
 перенаправляется на экран «Аудиоплеер».

HUEY-news committed:
1. Для всех TextView в макете activity_player
выставлен параметр android:lines="1"

HUEY-news committed:
1. Макет перепривязан снизу вверх.

HUEY-news committed:
1. Отформатирована дата в инфо о треке.

HUEY-news committed:
1. Форматирование даты перенесено в модель трека.

faritowich commented on app/src/main/java/com/practicum/playlistmaker/activity/PlayerActivity.kt:
```diff
@@ -0,0 +1,61 @@
+import java.text.SimpleDateFormat
+import java.util.Date
+
+const val TRACK_ID = "TRACK_ID"
```
Лучше хранить эту константу в companion object активити, где данные отправляются или принимаются. Чтобы константа не находилась сама по себе, без привязки к классам, где она используется

HUEY-news replied:
**Исправлено**

faritowich commented on app/src/main/java/com/practicum/playlistmaker/activity/PlayerActivity.kt:
```diff
@@ -0,0 +1,61 @@
+        val trackCountry = findViewById<TextView>(R.id.text_view_track_info_country_content)
+
+        val track: Track?
+        val json = intent.getStringExtra(TRACK_ID)
```
Текущая реализация небезопасная, так как если сейчас случайно передать из SearchActivity любой другой класс, то всё будет работать до  момента перехода на этот экран, после чего приложение упадет в рантайме. На самом деле можно передавать сам класс Track. Достаточно сделать его Serializable/Parcelable. Рекомендую почитать про это документацию, обычно свои классы аргументы передаются именно так.

HUEY-news replied:
Теперь минимальный api = **TIRAMISU**. Так как обычный метод getPercelableExtra с одним аргументом для интента стал **depricated**, а его перегруженная версия требует аннотации.

HUEY-news committed:
1. Константа TRACK_ID вынесена в companion object активити.

HUEY-news committed:
1. Класс Track теперь Parcelable.
2. Переписаны интенты вызова PlayerActivity.
3. Возникла проблема с десериализацией объекта Track.

HUEY-news committed:
1. Класс Track теперь Parcelable.
2. Переписаны интенты вызова PlayerActivity.
3. Возникла проблема с десериализацией объекта Track.
4. Минимальный API теперь TIRAMISU.
